test/dvotingdela.go

```diff
@@ -0,0 +1,519 @@
+	"go.dedis.ch/dela/core/ordering/cosipbft/authority"
+	"go.dedis.ch/dela/core/ordering/cosipbft/blockstore"
+	"go.dedis.ch/dela/core/ordering/cosipbft/types"
+	"go.dedis.ch/dela/core/store/hashtree"
+	"go.dedis.ch/dela/core/store/hashtree/binprefix"
+	"go.dedis.ch/dela/core/store/kv"
+	"go.dedis.ch/dela/core/txn"
+	"go.dedis.ch/dela/core/txn/pool"
+	poolimpl "go.dedis.ch/dela/core/txn/pool/gossip"
+	"go.dedis.ch/dela/core/txn/signed"
+	"go.dedis.ch/dela/core/validation"
+	"go.dedis.ch/dela/core/validation/simple"
+	"go.dedis.ch/dela/cosi/threshold"
+	"go.dedis.ch/dela/crypto"
+	"go.dedis.ch/dela/crypto/bls"
+	"go.dedis.ch/dela/crypto/loader"
+	"go.dedis.ch/dela/mino"
+	"go.dedis.ch/dela/mino/gossip"
+	"go.dedis.ch/dela/mino/minogrpc"
+	"go.dedis.ch/dela/mino/minogrpc/certs"
+	"go.dedis.ch/dela/mino/minogrpc/session"
+	"go.dedis.ch/dela/mino/router/tree"
+	"go.dedis.ch/dela/serde/json"
+	"golang.org/x/xerrors"
+)
+
+const certKeyName = "cert.key"
+const privateKeyFile = "private.key"
+
+var aKey = [32]byte{1}
+var valueAccessKey = [32]byte{2}
+
+// evotingAccessKey is the access key used for the evoting contract.
+var evotingAccessKey = [32]byte{3}
+
+// dela defines the common interface for a Dela node.
+type dela interface {
+	Setup(...dela)
+	GetMino() mino.Mino
+	GetOrdering() ordering.Service
+	GetTxManager() txn.Manager
+	GetAccessService() access.Service
+}
+
+// dVotingCosiDela defines the interface needed to use a Dela node using cosi.
+type dVotingCosiDela interface {
+	dela
+
+	GetAddr() net.Addr
+	GetPublicKey() crypto.PublicKey
+	GetPool() pool.Pool
+	GetAccessStore() accessstore
+	GetTree() hashtree.Tree
+	GetDkg() dkg.DKG
+	GetShuffle() shuffle.Shuffle
+
+	// jean ->
+	// adminID   string
+	// manager   *signed.TransactionManager
+	// firstNode dVotingNode
+	// actor     dkg.Actor
+	// signer    crypto.AggregateSigner
+	// <- jean
+}
+
+// dVotingNode represents a Dela node using cosi pbft aimed to execute d-voting
+//
+// - implements dela
+type dVotingNode struct {
+	t             *testing.T
+	onet          mino.Mino
+	ordering      ordering.Service
+	cosi          *threshold.Threshold
+	txManager     txn.Manager
+	addr          net.Addr
+	pool          pool.Pool
+	accessService access.Service
+	accessStore   accessstore
+	tree          hashtree.Tree
+	dkg           *pedersen.Pedersen
+	shuffle       *neff.NeffShuffle
+}
+
+// Creates n dela nodes using tempDir as root to file path and returns an array
+// of nodes or error
+func setupDVotingNodes(t *testing.T, numberOfNodes int, tempDir string) []dVotingCosiDela {
+	var dVotingNodes []dVotingCosiDela = make([]dVotingCosiDela, numberOfNodes)
+	var delaNodes []dela = make([]dela, numberOfNodes)
+
+	canal := make(chan dVotingCosiDela, numberOfNodes)
+
+	for i := 0; i < numberOfNodes; i++ {
+		filePath := filepath.Join(tempDir, "node", fmt.Sprint(i))
+		nodePort := 2000 + i
+
+		go newDVotingNode(canal, t, filePath, nodePort)
+		dVotingNodes[i] = <-canal
```

Comment:
This operation is blocking so you are not gaining anything. What you probably want to do is something like that:

```go
wait := sync.WaitGroup{}
wait.Add(numberOfNodes)

nodes := make(chan dVotingCosiDela, numberOfNodes)

for i := 0; i < numberOfNodes; i++ {
  go func(i int) {
    defer wait.Done()
    nodePort := 2000 + i
    nodes <- newDVotingNode(t, filePath, nodePort)
  } (i)
}

wait.Wait()
close(nodes)

dVotingNodes := make([]dVotingCosiDela, 0, numberOfNodes)

for node := range nodes {
  dVotingNodes = append(dVotingNodes, node)
}
```

Reply:
shouldn't we close(nodes) after the range loop ?

Reply:
No, otherwise it will loop forever. If I remember correctly once we close a channel we can't add new elements to that channel but we can still dry out its content.